File: GuardiasSaludMental/src/guardiassaludmental/guardias.js
/*
def is_valid_state(state):
    # check if it is a valid solution
    return True

def get_candidates(state):
    return []

def search(state, solutions):
    if is_valid_state(state):
        solutions.append(state.copy())
        # return

    for candidate in get_candidates(state):
        state.add(candidate)
        search(state, solutions)
        state.remove(candidate)

def solve():
    solutions = []
    state = set()
    search(state, solutions)
    return solutions

/*
 * @param {number[]} nums
 * @return {number[][]}
*/

/*
function subsets(nums) {
    let list = [];
    backtrack(list, [], nums, 0);
    return list;
}

function backtrack(list , tempList, nums, start){
	// cloning an array 
    // (ES6) - [...tempList] 
	// (ES5) - JSON.parse(JSON.stringify(tempList)) ;   
    list.push([...tempList]);
    for(let i = start; i < nums.length; i++){
        tempList.push(nums[i]);                   // choose
        backtrack(list, tempList, nums, i + 1);   // explore
        tempList.pop();                           // unchoose
    }
}
*/
/*
function nGuardias(n){
    let guardias = [];
    asignarGuardia(guardias, [], n, 0);
    
    console.log(`guardias: ${guardias}`);
    return guardias;
}
function asignarGuardia(guardias,candidatos,n,inicio){
    guardias.push(...candidatos);

    for(let i = inicio; i < n.length; i++){
        candidatos.push(n[i]);                   // choose
        asignarGuardia(guardias, candidatos, n, i + 1);   // explore
        candidatos.pop();                           // unchoose
    }

    
console.log(`candidatos: ${candidatos}`);
}

console.log(nGuardias(64));
*/

    
    // FUNCION 1 DEL BACKTRACKING

    function nGuardia(n){
        const docs = [{ 
            id:0,
            nombre:"Hector",
            dispDia:"LMXJVSD",
            dispNoche: "LMXJVSD",
            trabaja24: true,
            comienzoLicencia: "2023-07-01",
            finLicencia: "2023-07-15"
            },{
            id:1,
            nombre: "Maru",
            dispDia: "LV",
            dispNoche: "SD",
            trabaja24: false,
            comienzoLicencia: "2023-09-01",
            finLicencia: "2023-09-10"
            },{
            id:2,
            nombre: "Lau S.",
            dispDia: "LMXJVSD",
            dispNoche: "MS",
            trabaja24: true,
            comienzoLicencia: "2023-09-01",
            finLicencia: "2023-09-10"
            },{
            id:3,
            nombre: "Rodo",
            dispDia: "LMXJVSD",
            dispNoche: "LMXJVSD",
            trabaja24: true,
            comienzoLicencia: "2023-06-26",
            finLicencia: "2023-07-15"
            },{
            id:4,
            nombre: "Javier",
            dispDia: "LMXJSD",
            dispNoche: "LMSD",
            trabaja24: true,
            comienzoLicencia: "2023-06-16",
            finLicencia: "2023-06-21"
            },{
            id:5,
            nombre: "Flor",
            dispDia: "LMXJSD",
            dispNoche: "LMSD",
            trabaja24: false,
            comienzoLicencia: "2023-06-16",
            finLicencia: "2023-06-22"
            },{
            id:6,
            nombre: "Luisi",
            dispDia: "LXJSD",
            dispNoche: "D",
            trabaja24: false,
            comienzoLicencia: "2023-06-16",
            finLicencia: "2023-07-01"
            },{
            id:7,
            nombre: "Lucia",
            dispDia: "LVSD",
            dispNoche: "LMJSD",
            trabaja24: false,
            comienzoLicencia: "2023-09-16",
            finLicencia: "2023-10-01"
            },{
            id:8,
            nombre: "Lorena",
            dispDia: "MVS",
            dispNoche: "",
            trabaja24: false,
            comienzoLicencia: "2023-09-16",
            finLicencia: "2023-10-01"
            },{
            id:9,
            nombre: "Julia",
            dispDia: "VSD",
            dispNoche: "MXSD",
            trabaja24: true,
            comienzoLicencia: "2023-09-16",
            finLicencia: "2023-10-01"
            },{
            id:10,
            nombre: "Anto",
            dispDia: "LMXJVSD",
            dispNoche: "LMXJVSD",
            trabaja24: false,
            comienzoLicencia: "2023-09-16",
            finLicencia: "2023-10-01"
            }];
        const guardiaInicio = new Date("2023-06-15");
        const diaSemana = guardiaInicio.getDay();
        const nombreDia = ["D", "L", "M", "X", "J", "V", "S"];
        const guardia = {id:0, fecha: new Date(guardiaInicio), turno: "dia", dia: nombreDia[diaSemana], doc: " "};
    
        return asignarGuardia(docs,guardia,60)
    }

    function asignarGuardia(docs,guardia,n){
        let guardias = [];
        let candidatos = [];
        function recursive(docs,i){
            if(i===docs.length){
                return resultado.push([...candidatos]);
            }
            candidatos.push(guardias[i]);
            recursive(docs,guardias,i+1)

            candidatos.pop();
            recursive(guardias,i+1);
        }
        recursive(guardias,0)
    }
//     let guardias = 
//     let guardia = guardiaPrimera;
    
//     // console.log(guardias);

//     // La función recursiva se llama desde una función que no es recursiva y desde sí misma
//     return asignarGuardia(guardia,guardia,n);
// }

// FUNCION 2 DEL BACKTRACKING
function asignarGuardia(inicio,guardia,fin) {
    if (fin < inicio) {
        
        return {};
    } else {
        guardia=siguienteGuardia(guardia);
        const guardias = asignarGuardia(inicio,guardia,fin-1);
        console.log(guardias);
        for (const doc of docs) {
            if (esValido(guardia, doc)) {
                guardia.doc = doc.nombre;
                guardias[fin] = guardia;
                
            }
        }
        return guardias;
    }
}

// FUNCION 3 DEL BACKTRACKING
function esValido(guardia, doc) {
    let tieneDispo = 
        guardia.turno === "dia" ? 
            dispDia(doc, guardia.dia) 
            : 
            dispNoche(doc, guardia.dia);

        if (!tieneDispo) {
            return false;
        }
    return true;
}

function dispDia(doc, dia) {
    return doc.dispDia.includes(dia);
}

function dispNoche(doc, dia) {
    return doc.dispNoche.includes(dia);
}

function siguienteGuardia(guardia) {
    let siguienteFecha = new Date(guardia.fecha);

    if (guardia.turno === "noche") {
        siguienteFecha.setDate(siguienteFecha.getDate() + 1);
    }

    let siguienteTurno = guardia.turno === "dia" ? "noche" : "dia";
    let siguienteId = guardia.id + 1;

    let diaSemana = guardia.fecha.getDay();
    let nombreDia = ["D", "L", "M", "X", "J", "V", "S"];

    guardia = {id:siguienteId, fecha: siguienteFecha, turno: siguienteTurno, dia: nombreDia[diaSemana], doc: " "};
    
    return guardia;
}

asignarGuardia(1,guardiaPrimera,60);

File: GuardiasSaludMental/src/guardiassaludmental/guardias.py
# def esValido(guardia, doc, guardias):
#     # Check if the doctor was assigned to the previous shift
#     if guardia["id"] > 0 and guardias[guardia["id"] - 1] == doc["name"]:
#         print(f"Guard {doc['name']} cannot be assigned on {guardia['fecha']} ({guardia['turno']}) due to consecutive assignment.")
#         return False

#     # Check for guard availability
#     tiene_dispo = dispDia(doc, guardia["dia"]) if guardia["turno"] == "dia" else dispNoche(doc, guardia["dia"])
#     if not tiene_dispo:
#         print(f"Guard {doc['name']} is not available on {guardia['fecha']} ({guardia['turno']}).")
#         return False
    
#     # Check for guard leave days
#     if doc["comienzoLicencia"] <= guardia["fecha"] <= doc["finLicencia"]:
#         print(f"Guard {doc['name']} is on leave on {guardia['fecha']} and cannot be assigned.")
#         return False

#     return True

# # Run the function again to see the debugging output
# result_debug = nGuardias(30)
# result_debug

import random

# Generate a list of 10 doctors with random attribute values
docs = [{
        "name": "Hector",
        "dispDia": "LMXJVSD",
        "dispNoche": "LMXJVSD",
        "trabaja24": True,
        "comienzoLicencia": "2023-07-01",
        "finLicencia": "2023-07-15"
    },{
        "name": "Maru",
        "dispDia": "LV",
        "dispNoche": "SD",
        "trabaja24": False,
        "comienzoLicencia": "2023-09-01",
        "finLicencia": "2023-09-10"
    },{
        "name": "Lau S.",
        "dispDia": "LMXJVSD",
        "dispNoche": "MS",
        "trabaja24": True,
        "comienzoLicencia": "2023-09-01",
        "finLicencia": "2023-09-10"
    },{
        "name": "Rodo",
        "dispDia": "LMXJVSD",
        "dispNoche": "LMXJVSD",
        "trabaja24": True,
        "comienzoLicencia": "2023-06-26",
        "finLicencia": "2023-07-15"
    },{
        "name": "Javier",
        "dispDia": "LMXJSD",
        "dispNoche": "LMSD",
        "trabaja24": True,
        "comienzoLicencia": "2023-06-16",
        "finLicencia": "2023-06-21"
    },{
        "name": "Flor",
        "dispDia": "LMXJSD",
        "dispNoche": "LMSD",
        "trabaja24": False,
        "comienzoLicencia": "2023-06-16",
        "finLicencia": "2023-06-22"
    },{
        "name": "Luisi",
        "dispDia": "LXJSD",
        "dispNoche": "D",
        "trabaja24": False,
        "comienzoLicencia": "2023-06-16",
        "finLicencia": "2023-07-01"
    },{
        "name": "Lucia",
        "dispDia": "LVSD",
        "dispNoche": "LMJSD",
        "trabaja24": False,
        "comienzoLicencia": "2023-09-16",
        "finLicencia": "2023-10-01"
    },{
        "name": "Lorena",
        "dispDia": "MVS",
        "dispNoche": "",
        "trabaja24": False,
        "comienzoLicencia": "2023-09-16",
        "finLicencia": "2023-10-01"
    },{
        "name": "Julia",
        "dispDia": "VSD",
        "dispNoche": "MXSD",
        "trabaja24": True,
        "comienzoLicencia": "2023-09-16",
        "finLicencia": "2023-10-01"
    },{
        "name": "Anto",
        "dispDia": "LMXJVSD",
        "dispNoche": "LMXJVSD",
        "trabaja24": False,
        "comienzoLicencia": "2023-09-16",
        "finLicencia": "2023-10-01"
    }]

from datetime import datetime, timedelta

def nGuardias(n):
    guardia_inicio = datetime.strptime("2023-08-22", "%Y-%m-%d")
    dia_semana = guardia_inicio.weekday()
    nombre_dia = ["D", "L", "M", "X", "J", "V", "S"]
    guardia = {"id": 0, "fecha": guardia_inicio, "turno": "dia", "dia": nombre_dia[dia_semana]}

    guardias = [None] * n
    
    print(guardias)
    
    return asignarGuardia(guardia, guardias, n)

# Adding a dictionary to keep track of holiday shifts assigned to each doctor
holiday_shifts = {doc["name"]: 0 for doc in docs}



def esValido(guardia, doc, guardias):
    # Check if the doctor was assigned to the previous shift, and the current shift is a day shift (not allowed after a night shift)
    if guardia["id"] > 0 and guardias[guardia["id"] - 1] == doc["name"] and guardia["turno"] == "dia":
        print(f"Doctor {doc['name']} cannot be assigned on {guardia['fecha']} ({guardia['turno']}) due to consecutive assignment.")
        return False

    # Check for doctor availability
    tiene_dispo = dispDia(doc, guardia["dia"]) if guardia["turno"] == "dia" else dispNoche(doc, guardia["dia"])
    if not tiene_dispo:
        print(f"Doctor {doc['name']} is not available on {guardia['fecha']} ({guardia['turno']}).")
        return False
    
    # Check for doctor leave days
    if doc["comienzoLicencia"] <= guardia["fecha"] <= doc["finLicencia"]:
        print(f"Doctor {doc['name']} is on leave on {guardia['fecha']} and cannot be assigned.")
        return False

    # Check for equitable holiday shifts
    if guardia["dia"] == "D" and holiday_shifts[doc["name"]] >= (30/7) / len(docs):  # Assuming 30 days, 4 Sundays, divided by number of doctors
        print(f"Doctor {doc['name']} has already been assigned the maximum number of holiday shifts.")
        return False

    return True

# Adjusting the assignment function to update the holiday shifts count
def asignarGuardia(guardia, guardias, n):
    if guardia["id"] == (n - 1):
        return guardias
    else:
        total_sol = 0
        for doc in docs:
            if esValido(guardia, doc, guardias):
                guardias[guardia["id"]] = doc["name"]
                if guardia["dia"] == "D":
                    holiday_shifts[doc["name"]] += 1
                guardia_siguiente = siguienteGuardia(guardia)
                total_sol += asignarGuardia(guardia_siguiente, guardias, n)
        return total_sol

def dispDia(doc, dia):
    return dia in doc["dispDia"]

def dispNoche(doc, dia):
    return dia in doc["dispNoche"]

def siguienteGuardia(guardia):
    siguiente_fecha = guardia["fecha"]

    if guardia["turno"] == "noche":
        siguiente_fecha += timedelta(days=1)

    siguiente_turno = "noche" if guardia["turno"] == "dia" else "dia"
    siguiente_id = guardia["id"] + 1

    return {"id": siguiente_id, "fecha": siguiente_fecha, "turno": siguiente_turno}

# Run the function again with the new logic
result_debug = nGuardias(30)
result_debug

# Run the function with the new set of doctors
result_debug = asignarGuardia({"id": 0, "fecha": "2023-08-01", "dia": "D", "turno": "dia"}, [""] * 30, 30)
print(result_debug)